### PR TITLE
usecases: Future state machine client

### PR DESCRIPTION
**Pull request by j-mueller**
* Use the state machine client lib. in the futures contract
* We want to avoid computing the actual validator
  validator script of the contract too often. The approach is similar to what
  was done in the `MultiSigStateMachine` contract: Re-use the
  contract's `ScriptInstance` as much as possible. This is
  achieved by sharing the script instance value instead of computing it
  from scratch, and by checkpointing it after it has been created.
* Test times are now down from ~30s to 5-8s - still not perfect but
  tolerable IMO.

### CODE REVIEW

File: plutus-use-cases/bench/Bench.hs [hunk 1]
```diff
@@ -284,3 +291,36 @@ mockPendingTx = PendingTx
...
+
+unValidatorHash :: ValidatorScript -> PlutusTx.ByteString
+unValidatorHash vs = 
+    let Ledger.Scripts.ValidatorHash h = Ledger.Scripts.validatorHash vs
+    in h
+
+-- Script hashes
+scriptHashes :: Benchmark
+scriptHashes = bgroup "script hashes" [
+    let si = TA.scriptInstance (TA.Account ("fd2c8c0705d3ca1e7b1aeaa4da85dfe5ac6dde64da9d241011d84c0ee97aac5e", "my token")) in
```
Comment (michaelpj) [inline plutus-use-cases/bench/Bench.hs]: hardcoded hashes :crying_cat_face:

File: plutus-use-cases/bench/Bench.hs [hunk 2]
```diff
@@ -284,3 +291,36 @@ mockPendingTx = PendingTx
     , pendingTxId = TxId P.emptyByteString
     , pendingTxData = []
     }
+
+unValidatorHash :: ValidatorScript -> PlutusTx.ByteString
```
Comment (michaelpj) [inline plutus-use-cases/bench/Bench.hs]: This seems named wrong.

Comment (michaelpj) [reply plutus-use-cases/bench/Bench.hs]: And unused. Maybe it was to get rid of the hardcoded hashes?

File: plutus-use-cases/plutus-use-cases.cabal [hunk 3]
```diff
@@ -132,6 +132,8 @@ benchmark plutus-use-cases-bench
     ghc-options: -Wall -Wincomplete-uni-patterns
                  -Wincomplete-record-updates -Wredundant-constraints -Widentities
                  -rtsopts
+    build-depends:
+        plutus-emulator -any
```
Comment (michaelpj) [inline plutus-use-cases/plutus-use-cases.cabal]: Can't it be in the main `build-depends`?

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Escrow.hs [hunk 4]
```diff
@@ -234,23 +228,25 @@ payEp
...
 -- | Pay some money into the escrow contract.
 pay
     :: ( HasWriteTx s
        , HasOwnPubKey s
        , AsContractError e
        )
-    => EscrowParams DataValue
+    => ScriptInstance Escrow
+    -- ^ The instance
+    -> EscrowParams DataValue
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Escrow.hs]: I do find it a bit gross to be passing both the params and the instance derived from them, but I'm not sure what else we can do here.

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs [hunk 5]
```diff
@@ -76,36 +73,103 @@ import qualified Prelude as Haskell
...
+    --   *didn't* violate the margin requirement
+    deriving (Show)
+
 data FutureError =
     TokenSetupFailed ContractError
     -- ^ Something went wrong during the setup of the two tokens
-    | MarginAdjustmentFailed ContractError
+    | StateMachineError (SM.SMContractError FutureState FutureAction)
+    | OtherFutureError ContractError
     -- ^ Something went wrong during a margin payment
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs]: Wrong comment

File: plutus-wallet-api/ledger/Ledger/Typed/Scripts.hs [hunk 6]
```diff
@@ -42,35 +46,32 @@ type ValidatorType (a :: Type) = DataType a -> RedeemerType a -> Validation.Pend
...
-data ScriptInstance (a :: Type) where
-    Validator ::
-        ScriptType a
-        => CompiledCode (ValidatorType a)
-        -> CompiledCode (ValidatorType a -> WrappedValidatorType)
-        -> Validator
-        -> Addr.Address
-        -> ScriptInstance a
+-- | A typed validator script with its 'ValidatorScript' and 'Address'.
+data ScriptInstance (a :: Type) =
```
Comment (michaelpj) [inline plutus-wallet-api/ledger/Ledger/Typed/Scripts.hs]: Why the change here?

Comment (j-mueller) [reply plutus-wallet-api/ledger/Ledger/Typed/Scripts.hs]: We don't use the `CompiledCode` values anywhere (also I wanted to get the JSON instances so that I can save the instance with a `checkpoint` in the future contract)

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Escrow.hs (hunk 4 above)
Comment (j-mueller) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Escrow.hs]: I can add a type that combines the two, but I don't think there's a way around passing the two of them around together if we want to avoid computing the instance over and over again (which causes the slowdown)

File: plutus-contract/src/Language/Plutus/Contract/StateMachine.hs [hunk 7]
```diff
@@ -114,7 +114,7 @@ defaultChooser xs  =
 mkStateMachineClient ::
     forall state input
     . SM.StateMachineInstance state input
-    -> (state -> input -> Value -> ValueAllocation)
+    -> (state -> input -> Value -> Maybe ValueAllocation)
```
[the remark this answers is not in the document]
Comment (j-mueller) [reply plutus-contract/src/Language/Plutus/Contract/StateMachine.hs]: `smCheck` is going to be replaced by this (in #1753), and then we definitely need the `Maybe`.

File: plutus-use-cases/bench/Bench.hs (hunk 1 above)
Comment (j-mueller) [reply plutus-use-cases/bench/Bench.hs]: in this instance the value itself doesn't matter, so at least this won't break when the contracts change (small consolation, I know)

File: plutus-wallet-api/ledger/Ledger/Typed/Scripts.hs (hunk 6 above)
Comment (michaelpj) [reply plutus-wallet-api/ledger/Ledger/Typed/Scripts.hs]: Hmm okay. I guess it's just a phantom type now, which seems fine.
